### PR TITLE
Add callbacks for scanning evaluation context

### PR DESCRIPTION
**Pull request by jonatanklosko**
See https://github.com/livebook-dev/livebook/pull/1041.

### CODE REVIEW

File: lib/kino/smart_cell.ex [hunk 1]
```diff
@@ -160,6 +160,31 @@ defmodule Kino.SmartCell do
...
+  at variables, instead of returning their values, extract and return
+  only the relevant metadata.
+  """
+  @callback scan_binding(Code.binding(), Macro.Env.t()) :: data :: term()
+
+  @doc """
+  Invoked with the result of `c:scan_binding/2` to update the server
+  state.
+  """
+  @callback scan_binding_continue(data :: term(), ctx :: Context.t()) :: ctx :: Context.t()
```
Comment (jonatanklosko) [inline lib/kino/smart_cell.ex:184]: @josevalim we could similarly add `scan_evaluation_result` and `scan_evaluation_result_continue`. Implementation-wise I think it's pretty much the same as `handle_exception`. It's more verbose, but also gives more flexibility. Thoughts?

Comment (jonatanklosko) [reply lib/kino/smart_cell.ex:184]: To only handle errors we could then do:

```elixir
@impl true
def scan_evaluation_result({:ok, _}), do: :none
def scan_evaluation_result({:error, kind, reason, stacktrace}), do: {kind, reason, stacktrace}

@impl true
def scan_evaluation_result_continue(exception) do
  # handle_exception
end
```

File: lib/kino/smart_cell.ex [hunk 2]
```diff
@@ -160,6 +160,34 @@ defmodule Kino.SmartCell do
...
+  **Important:** avoid any heavy work in this callback, as it runs in
+  the same process that evaluates code, so we don't want to block it.
+  """
+  @callback scan_binding(Code.binding(), Macro.Env.t()) :: data :: term()
+
+  @doc """
+  Invoked with the result of `c:scan_binding/2` to update the server
+  state.
+  """
+  @callback scan_binding_continue(data :: term(), ctx :: Context.t()) :: ctx :: Context.t()
```
Comment (josevalim) [inline lib/kino/smart_cell.ex:187]: For completeness, there is another API you can use here:

    scan_binding(pid, code, env)

Where `pid` is the `pid` of the smart cell. This way we have fewer callbacks and give a bit more flexibility to the user. I think I prefer this approach.

If you prefer two callbacks, then i would perhaps change the names to: `prepare_scan_binding` and `handle_scan_binding`. `handle_` is a known name for server callbacks after all and that would make it more apparent.

**Review by josevalim — APPROVED** [review]
LGTM! Just one tiny comment!

**Review by josevalim — APPROVED** [review]

**Pull request merged**